### PR TITLE
Merge Mining Proxy - Update

### PR DESCRIPTION
**Pull request by StriderDM**
## Description
Improved logging
Added default logging for merge mining proxy to log4rs-sample
Fixed most clippy lints
Added height check for tari chain

## How Has This Been Tested?
cargo test --bin tari_merge_mining_proxy
Manually run

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [x] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

### CODE REVIEW

File: applications/tari_merge_mining_proxy/Cargo.toml [hunk 1]
```diff
@@ -33,6 +33,7 @@ rand = "0.7.2"
 derive-error = "0.0.4"
 thiserror = "1.0.15"
 structopt = { version = "0.3.13", default_features = false }
+levenshtein-rs = "0.1.0"
```
Comment (SWvheerden) [inline applications/tari_merge_mining_proxy/Cargo.toml]: I would leave this out till we started using it.

File: applications/tari_merge_mining_proxy/src/main.rs [hunk 2]
```diff
@@ -66,7 +66,7 @@ use tokio::runtime::Runtime;
 
 pub type PrivateKey = RistrettoSecretKey;
 
-pub const LOG_TARGET: &str = "base_node::conduit_app";
+pub const LOG_TARGET: &str = "tari_mm_proxy";
```
Comment (SWvheerden) [inline applications/tari_merge_mining_proxy/src/main.rs]:
```suggestion
pub const LOG_TARGET: &str = "tari_mm_proxy::main";
```
We might have more than one file in the logging of the application later.

File: applications/tari_merge_mining_proxy/src/main.rs [hunk 3]
```diff
@@ -340,14 +363,47 @@ fn handle_connection(
         .map_err(|e| MmProxyError::OtherError(format!("Error reading from stream {}", e)))?;
     let date = Local::now();
     let request_string = stringify_request(&buffer[..]);
+    debug!(target: LOG_TARGET, "Request: {}", request_string);
     let request_type = get_request_type(&buffer[..])?;
     let url_part = get_url_part(&buffer[..])?;
     if request_type.starts_with("GET") {
-        debug!(target: LOG_TARGET, "Handling Method: {}", url_part);
+        info!(target: LOG_TARGET, "Handling GET Method: {}", url_part);
```
Comment (SWvheerden) [inline applications/tari_merge_mining_proxy/src/main.rs]:
```suggestion
        debug!(target: LOG_TARGET, "Handling GET Method: {}", url_part);
```
This should not be info
this should be debug. 

Info messages are messages that the non technical user might want to see, like accepted new block, mined new block. 
Not handling metod etc.

File: applications/tari_merge_mining_proxy/src/main.rs [hunk 4]
```diff
@@ -340,14 +363,47 @@ fn handle_connection(
         .map_err(|e| MmProxyError::OtherError(format!("Error reading from stream {}", e)))?;
     let date = Local::now();
     let request_string = stringify_request(&buffer[..]);
+    debug!(target: LOG_TARGET, "Request: {}", request_string);
```
Comment (SWvheerden) [inline applications/tari_merge_mining_proxy/src/main.rs:366]:
```suggestion
    trace!(target: LOG_TARGET, "Request: {}", request_string);
```

File: applications/tari_merge_mining_proxy/src/main.rs [hunk 5]
```diff
@@ -340,14 +363,47 @@ fn handle_connection(
...
+                            }
+                            transient.tari_height = Some(height);
+                        },
+                        Err(e) => {
+                            debug!(target: LOG_TARGET, "{:#}", e);
+                        },
+                    }
+                },
+                Err(e) => {
+                    debug!(target: LOG_TARGET, "{:#}", e);
```
Comment (SWvheerden) [inline applications/tari_merge_mining_proxy/src/main.rs]:
```suggestion
                    error!(target: LOG_TARGET, "{:#}", e);
```

File: applications/tari_merge_mining_proxy/src/main.rs [hunk 6]
```diff
@@ -340,14 +363,47 @@ fn handle_connection(
...
+                                    if submit_height >= current_height {
+                                        adjust_height(&data, current_height)?;
+                                        info!(target: LOG_TARGET, "Already submitted for current Tari height");
+                                    }
+                                }
+                            }
+                            transient.tari_height = Some(height);
+                        },
+                        Err(e) => {
+                            debug!(target: LOG_TARGET, "{:#}", e);
```
Comment (SWvheerden) [inline applications/tari_merge_mining_proxy/src/main.rs]:
```suggestion
                            error!(target: LOG_TARGET, "{:#}", e);
```

File: applications/tari_merge_mining_proxy/src/main.rs [hunk 7]
```diff
@@ -340,14 +363,47 @@ fn handle_connection(
...
+                            // Short Circuit XMRig to request a new block template
+                            // TODO: needs additional testing
+                            if let Some(current_height) = transient.tari_height {
+                                if height != current_height {
+                                    adjust_height(&data, current_height)?;
+                                    info!(target: LOG_TARGET, "Tari tip changed");
+                                } else if let Some(submit_height) = transient.tari_prev_submit_height {
+                                    if submit_height >= current_height {
+                                        adjust_height(&data, current_height)?;
+                                        info!(target: LOG_TARGET, "Already submitted for current Tari height");
```
Comment (SWvheerden) [inline applications/tari_merge_mining_proxy/src/main.rs]:
```suggestion
                                        debug!(target: LOG_TARGET, "Already submitted for current Tari height");
```
Non technical user I dont think wants to see this

File: applications/tari_merge_mining_proxy/src/main.rs [hunk 8]
```diff
@@ -340,14 +363,47 @@ fn handle_connection(
...
+                Ok(mut grpcclient) => {
+                    match rt.block_on(grpcclient.get_tip_info(grpc::Empty {})) {
+                        Ok(result) => {
+                            let height = result.into_inner().metadata.unwrap().height_of_longest_chain;
+                            // Short Circuit XMRig to request a new block template
+                            // TODO: needs additional testing
+                            if let Some(current_height) = transient.tari_height {
+                                if height != current_height {
+                                    adjust_height(&data, current_height)?;
+                                    info!(target: LOG_TARGET, "Tari tip changed");
```
Comment (SWvheerden) [inline applications/tari_merge_mining_proxy/src/main.rs]: This is debatable if its info/debug. 
My vote is slightly for debug.

File: applications/tari_merge_mining_proxy/src/main.rs [hunk 9]
```diff
@@ -361,13 +417,9 @@ fn handle_connection(
...
             let url = format!("{}{}", MONEROD_URL, url_part);
             let mut curl = base_curl(json.len() as u64, &url, true)?;
-            debug!(target: LOG_TARGET, "Request: {}", request_string);
             let method = get_post_json_method(&json)?;
             let mut data = do_curl(&mut curl, &json)?;
-
-            // TODO: Check tari height, monero check height request is type get (/getheight)
-
-            debug!(target: LOG_TARGET, "Handling Method: {}", method.as_str());
+            info!(target: LOG_TARGET, "Handling POST Method: {}", method.as_str());
```
Comment (SWvheerden) [inline applications/tari_merge_mining_proxy/src/main.rs]:
```suggestion
            debug!(target: LOG_TARGET, "Handling POST Method: {}", method.as_str());
```

File: applications/tari_merge_mining_proxy/src/main.rs [hunk 10]
```diff
@@ -392,14 +444,15 @@ fn handle_connection(
...
                                         tariheader.pow = Some(powdata);
                                         block.header = Some(tariheader);
                                         trace!(target: LOG_TARGET, "Tari Block {:?}", block);
                                         match rt.block_on(grpcclient.submit_block(block)) {
                                             Ok(result) => {
                                                 result.into_inner();
+                                                transient.tari_prev_submit_height = transient.tari_height;
                                             },
                                             Err(e) => {
                                                 debug!(target: LOG_TARGET, "{:#}", e);
```
Comment (SWvheerden) [inline applications/tari_merge_mining_proxy/src/main.rs]:
```suggestion
                                                error!(target: LOG_TARGET, "{:#}", e);
```

Comment (SWvheerden) [reply applications/tari_merge_mining_proxy/src/main.rs]: might be warn, I am not sure which result we are looking there at, the result of the grpc submit, or die actual submit block call. 

If its the grpc fail its error. 
If its the actual call, its a warn

File: applications/tari_merge_mining_proxy/src/main.rs [hunk 11]
```diff
@@ -456,7 +509,6 @@ fn handle_connection(
                                                 target: LOG_TARGET,
                                                 "BlockHashBlob: {:#}", &parsed["result"]["blockhashing_blob"]
                                             );
-                                            debug!(target: LOG_TARGET, "Transient: {:?}", transient);
                                         },
                                         Err(e) => {
                                             debug!(target: LOG_TARGET, "{:#}", e);
```
Comment (SWvheerden) [inline applications/tari_merge_mining_proxy/src/main.rs]: Same as above comment

**Review by SWvheerden — COMMENTED** [review]
Looks good. 
Some comments of the log levels. I see many if them are not correct. 
Log levels I see as the following, I think @mikethetike you might want to comment here:
Info: Console feedback for non technical users
Warn: Feedback about some potential problem, not app breaking error
Error: Some app error that inhibits functionality. 
Debug: Information for developers tracing a problem
Trace: Raw detailed tracing information

File: applications/tari_merge_mining_proxy/Cargo.toml (hunk 1 above)
Comment (StriderDM) [reply applications/tari_merge_mining_proxy/Cargo.toml]: Done

File: applications/tari_merge_mining_proxy/src/main.rs (hunk 3 above)
Comment (StriderDM) [reply applications/tari_merge_mining_proxy/src/main.rs]: Done

File: applications/tari_merge_mining_proxy/src/main.rs (hunk 2 above)
Comment (StriderDM) [reply applications/tari_merge_mining_proxy/src/main.rs]: Have updated this.

File: applications/tari_merge_mining_proxy/src/main.rs (hunk 4 above)
Comment (StriderDM) [reply applications/tari_merge_mining_proxy/src/main.rs:366]: This needs to be debug at least to keep track of what is going on.

File: applications/tari_merge_mining_proxy/src/main.rs (hunk 8 above)
Comment (StriderDM) [reply applications/tari_merge_mining_proxy/src/main.rs]: Changed to debug.

File: applications/tari_merge_mining_proxy/src/main.rs (hunk 6 above)
Comment (StriderDM) [reply applications/tari_merge_mining_proxy/src/main.rs]: Have updated instances of `Err(e)` to `error!`

File: applications/tari_merge_mining_proxy/src/main.rs [hunk 12]
```diff
@@ -456,21 +508,20 @@ fn handle_connection(
...
                                 Err(e) => {
-                                    debug!(target: LOG_TARGET, "{:#}", e);
+                                    error!(target: LOG_TARGET, "{:#}", e);
                                 },
                             }
                         },
                         Err(e) => {
-                            debug!(target: LOG_TARGET, "Tari base node unavailable, monero mining only");
-                            debug!(target: LOG_TARGET, "{:#}", e);
+                            error!(target: LOG_TARGET, "Tari base node unavailable, monero mining only");
```
Comment (stringhandler) [inline applications/tari_merge_mining_proxy/src/main.rs:523]: nit: These can probably be one message

**Review by stringhandler — DISMISSED** [review]
Looks good. (Untested)

**Pull request merged**